From 09d78eb1f1ed8b9fef8b421ba5bc132cc7e457c2 Mon Sep 17 00:00:00 2001
From: romain pelletant <romainp@kickmaker.net>
Date: Tue, 20 Dec 2022 19:04:46 +0100
Subject: [PATCH] net: lwm2m: add callback for send confirmation

Make send result deprecated as confirmation result shall always be asked.
Add user callback to get confirmation result. See #52328

Signed-off-by: romain pelletant <romainp@kickmaker.net>
---
 include/zephyr/net/lwm2m.h                    | 55 +++++++++++++++----
 subsys/net/lib/lwm2m/lwm2m_message_handling.c | 51 +++++++++++++----
 subsys/net/lib/lwm2m/lwm2m_object.h           |  1 +
 subsys/net/lib/lwm2m/lwm2m_shell.c            | 27 +++------
 4 files changed, 91 insertions(+), 43 deletions(-)

diff --git a/include/zephyr/net/lwm2m.h b/include/zephyr/net/lwm2m.h
index 9c5745c272..86691a1b35 100644
--- a/include/zephyr/net/lwm2m.h
+++ b/include/zephyr/net/lwm2m.h
@@ -2085,35 +2085,68 @@ void lwm2m_rd_client_update(void);
  */
 char *lwm2m_path_log_buf(char *buf, struct lwm2m_obj_path *path);
 
-/** 
- * @brief LwM2M SEND operation to given path list
+/**
+ * @brief LwM2M send status
  *
- * @deprecated Use lwm2m_send() instead.
+ * LwM2M send status are generated back to the lwm2m_send_cb_t function in
+ * lwm2m_send_cb()
+ */
+enum lwm2m_send_status {
+	LWM2M_SEND_STATUS_SUCCESS,
+	LWM2M_SEND_STATUS_FAILURE,
+	LWM2M_SEND_STATUS_TIMEOUT,
+};
+
+/**
+ * @typedef lwm2m_send_cb_t
+ * @brief Callback returning send status
+ */
+typedef void (*lwm2m_send_cb_t)(enum lwm2m_send_status status);
+
+/**
+ * @brief LwM2M SEND operation to given path list asynchronously with confirmation callback
  *
  * @param ctx LwM2M context
- * @param path_list LwM2M Path string list
+ * @param path_list LwM2M path struct list
  * @param path_list_size Length of path list. Max size is CONFIG_LWM2M_COMPOSITE_PATH_LIST_SIZE
- * @param confirmation_request True request confirmation for operation.
+ * @param confirmation_request Set to true to use CoAP CON request which is the only allowed.
+ *
+ * @return 0 for success or negative in case of error.
+ *
+ */
+int lwm2m_send(struct lwm2m_ctx *ctx, const struct lwm2m_obj_path path_list[],
+			  uint8_t path_list_size, bool confirmation_request);
+
+/** 
+ * @brief LwM2M SEND operation to given path list asynchronously with confirmation callback
+ *
+ * @param ctx LwM2M context
+ * @param path_list LwM2M path struct list
+ * @param path_list_size Length of path list. Max size is CONFIG_LWM2M_COMPOSITE_PATH_LIST_SIZE
+ * @param reply_cb Callback triggered with confirmation state or NULL if not used
  * 
  * @return 0 for success or negative in case of error.
  *
  */
-int lwm2m_engine_send(struct lwm2m_ctx *ctx, char const *path_list[], uint8_t path_list_size,
-		      bool confirmation_request);
+int lwm2m_send_cb(struct lwm2m_ctx *ctx, const struct lwm2m_obj_path path_list[],
+			  uint8_t path_list_size, lwm2m_send_cb_t reply_cb);
 
 /** 
- * @brief LwM2M SEND operation to given path list
+ * @brief LwM2M SEND operation to given path list asynchronously without confirmation
+ *
+ * @deprecated Use lwm2m_send_cb() instead, to check for confirmation request callbacks
  *
  * @param ctx LwM2M context
- * @param path_list LwM2M path struct list
+ * @param path_list LwM2M Path string list
  * @param path_list_size Length of path list. Max size is CONFIG_LWM2M_COMPOSITE_PATH_LIST_SIZE
  * @param confirmation_request True request confirmation for operation.
  * 
  * @return 0 for success or negative in case of error.
  *
  */
-int lwm2m_send(struct lwm2m_ctx *ctx, const struct lwm2m_obj_path path_list[],
-	       uint8_t path_list_size, bool confirmation_request);
+__deprecated
+int lwm2m_engine_send(struct lwm2m_ctx *ctx, char const *path_list[], uint8_t path_list_size,
+		      bool confirmation_request);
 
 /** 
  * @brief Returns LwM2M client context
diff --git a/subsys/net/lib/lwm2m/lwm2m_message_handling.c b/subsys/net/lib/lwm2m/lwm2m_message_handling.c
index 877ee75e10..09804018c8 100644
--- a/subsys/net/lib/lwm2m/lwm2m_message_handling.c
+++ b/subsys/net/lib/lwm2m/lwm2m_message_handling.c
@@ -2880,6 +2880,9 @@ static int do_send_reply_cb(const struct coap_packet *response, struct coap_repl
 			    const struct sockaddr *from)
 {
 	uint8_t code;
+	struct lwm2m_message *msg = NULL;
+
+	msg = (struct lwm2m_message *) reply->user_data;
 
 	code = coap_header_get_code(response);
 	LOG_DBG("Send callback (code:%u.%u)", COAP_RESPONSE_CODE_CLASS(code),
@@ -2887,17 +2890,27 @@ static int do_send_reply_cb(const struct coap_packet *response, struct coap_repl
 
 	if (code == COAP_RESPONSE_CODE_CHANGED) {
 		LOG_INF("Send done!");
+		if (msg && msg->send_status_cb) {
+			msg->send_status_cb(LWM2M_SEND_STATUS_SUCCESS);
+		}
 		return 0;
 	}
 
 	LOG_ERR("Failed with code %u.%u. Not Retrying.", COAP_RESPONSE_CODE_CLASS(code),
 		COAP_RESPONSE_CODE_DETAIL(code));
 
+	if (msg && msg->send_status_cb) {
+		msg->send_status_cb(LWM2M_SEND_STATUS_FAILURE);
+	}
+
 	return 0;
 }
 
 static void do_send_timeout_cb(struct lwm2m_message *msg)
 {
+	if (msg->send_status_cb) {
+		msg->send_status_cb(LWM2M_SEND_STATUS_TIMEOUT);
+	}
 	LOG_WRN("Send Timeout");
 	lwm2m_rd_client_timeout(msg->ctx);
 }
@@ -2938,7 +2951,17 @@ static bool init_next_pending_timeseries_data(struct lwm2m_cache_read_info *cach
 #endif
 
 int lwm2m_send(struct lwm2m_ctx *ctx, const struct lwm2m_obj_path path_list[],
-	       uint8_t path_list_size, bool confirmation_request)
+			 uint8_t path_list_size, bool confirmation_request)
+{
+	if (!confirmation_request) {
+		return -EINVAL;
+	}
+
+	return lwm2m_send_cb(ctx, path_list, path_list_size, NULL);
+}
+
+int lwm2m_send_cb(struct lwm2m_ctx *ctx, const struct lwm2m_obj_path path_list[],
+			 uint8_t path_list_size, lwm2m_send_cb_t reply_cb)
 {
 #if defined(CONFIG_LWM2M_SERVER_OBJECT_VERSION_1_1)
 	struct lwm2m_message *msg;
@@ -3007,15 +3030,10 @@ msg_alloc:
 
 msg_init:
 
-	if (confirmation_request) {
-		msg->type = COAP_TYPE_CON;
-		msg->reply_cb = do_send_reply_cb;
-		msg->message_timeout_cb = do_send_timeout_cb;
-	} else {
-		msg->type = COAP_TYPE_NON_CON;
-		msg->reply_cb = NULL;
-		msg->message_timeout_cb = NULL;
-	}
+	msg->type = COAP_TYPE_CON;
+	msg->reply_cb = do_send_reply_cb;
+	msg->message_timeout_cb = do_send_timeout_cb;
+
 	msg->code = COAP_METHOD_POST;
 	msg->mid = coap_next_id();
 	msg->tkl = LWM2M_MSG_TOKEN_GENERATE_NEW;
@@ -3029,6 +3047,12 @@ msg_init:
 	msg->cache_info = &cache_temp_info;
 #endif
 
+	/* Register user callback if defined for confirmation */
+	if (reply_cb) {
+		msg->reply->user_data = msg;
+		msg->send_status_cb = (lwm2m_send_cb_t) reply_cb;
+	}
+
 	ret = select_writer(&msg->out, content_format);
 	if (ret) {
 		goto cleanup;
@@ -3056,7 +3080,7 @@ msg_init:
 #if defined(CONFIG_LWM2M_RESOURCE_DATA_CACHE_SUPPORT)
 	msg->cache_info = NULL;
 #endif
-	LOG_INF("Send op to server (/dp)");
+	LOG_DBG("Send op to server (/dp)");
 	lwm2m_information_interface_send(msg);
 
 #if defined(CONFIG_LWM2M_RESOURCE_DATA_CACHE_SUPPORT)
@@ -3089,6 +3113,9 @@ int lwm2m_engine_send(struct lwm2m_ctx *ctx, char const *path_list[], uint8_t pa
 {
 	int ret;
 	struct lwm2m_obj_path lwm2m_path_list[CONFIG_LWM2M_COMPOSITE_PATH_LIST_SIZE];
+	if (!confirmation_request) {
+		return -EINVAL;
+	}
 
 	if (path_list_size > CONFIG_LWM2M_COMPOSITE_PATH_LIST_SIZE) {
 		return -E2BIG;
@@ -3102,5 +3129,5 @@ int lwm2m_engine_send(struct lwm2m_ctx *ctx, char const *path_list[], uint8_t pa
 		}
 	}
 
-	return lwm2m_send(ctx, lwm2m_path_list, path_list_size, confirmation_request);
+	return lwm2m_send_cb(ctx, lwm2m_path_list, path_list_size, NULL);
 }
diff --git a/subsys/net/lib/lwm2m/lwm2m_object.h b/subsys/net/lib/lwm2m/lwm2m_object.h
index f9def911f9..218568e37d 100644
--- a/subsys/net/lib/lwm2m/lwm2m_object.h
+++ b/subsys/net/lib/lwm2m/lwm2m_object.h
@@ -500,6 +500,7 @@ struct lwm2m_message {
 	uint8_t *token;
 	coap_reply_t reply_cb;
 	lwm2m_message_timeout_cb_t message_timeout_cb;
+	lwm2m_send_cb_t send_status_cb;
 	uint16_t mid;
 	uint8_t type;
 	uint8_t code;
diff --git a/subsys/net/lib/lwm2m/lwm2m_shell.c b/subsys/net/lib/lwm2m/lwm2m_shell.c
index 5677eb3043..92fd7e72b4 100644
--- a/subsys/net/lib/lwm2m/lwm2m_shell.c
+++ b/subsys/net/lib/lwm2m/lwm2m_shell.c
@@ -22,8 +22,7 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
 #include <lwm2m_util.h>
 
 #define LWM2M_HELP_CMD "LwM2M commands"
-#define LWM2M_HELP_SEND "LwM2M SEND operation\nsend [OPTION]... [PATH]...\n" \
-	"-n\t Send as non-confirmable\n" \
+#define LWM2M_HELP_SEND "LwM2M SEND operation\nsend [PATH]...\n" \
 	"Root-level operation is unsupported"
 #define LWM2M_HELP_EXEC "Execute a resource\nexec PATH\n"
 #define LWM2M_HELP_READ "Read value from LwM2M resource\nread PATH [OPTIONS]\n" \
@@ -60,8 +59,6 @@ static int cmd_send(const struct shell *sh, size_t argc, char **argv)
 	int ret = 0;
 	struct lwm2m_ctx *ctx = lwm2m_rd_client_ctx();
 	int path_cnt = argc - 1;
-	bool confirmable = true;
-	int ignore_cnt = 1; /* Subcmd + arguments preceding the path list */
 	struct lwm2m_obj_path lwm2m_path_list[CONFIG_LWM2M_COMPOSITE_PATH_LIST_SIZE];
 
 	if (!ctx) {
@@ -70,18 +67,6 @@ static int cmd_send(const struct shell *sh, size_t argc, char **argv)
 	}
 
 	if (argc < 2) {
-		shell_error(sh, "no arguments or path(s)\n");
-		shell_help(sh);
-		return -EINVAL;
-	}
-
-	if (strcmp(argv[1], "-n") == 0) {
-		confirmable = false;
-		path_cnt--;
-		ignore_cnt++;
-	}
-
-	if ((argc - ignore_cnt) == 0) {
 		shell_error(sh, "no path(s)\n");
 		shell_help(sh);
 		return -EINVAL;
@@ -91,18 +76,20 @@ static int cmd_send(const struct shell *sh, size_t argc, char **argv)
 		return -E2BIG;
 	}
 
-	for (int i = ignore_cnt; i < path_cnt; i++) {
+	for (int i = 0; i < path_cnt; i++) {
 		/* translate path -> path_obj */
-		ret = lwm2m_string_to_path(argv[i], &lwm2m_path_list[i], '/');
+		const char *p = argv[1 + i];
+
+		ret = lwm2m_string_to_path(p, &lwm2m_path_list[i], '/');
 		if (ret < 0) {
 			return ret;
 		}
 	}
 
-	ret = lwm2m_send(ctx, lwm2m_path_list, path_cnt, confirmable);
+	ret = lwm2m_send(ctx, lwm2m_path_list, path_cnt, true);
 
 	if (ret < 0) {
-		shell_error(sh, "can't do send operation, request failed\n");
+		shell_error(sh, "can't do send operation, request failed (%d)\n", ret);
 		return -ENOEXEC;
 	}
 	return 0;
-- 
2.34.1

